### PR TITLE
[Back+Front] Message Board

### LINKED ISSUE #109: create the base for Messages
## Messages

We need to make a base for the messages page for each course.

### PR DESCRIPTION
**Pull request by shakedhagag**
## MessageBoard
I have implemented the message board model:
1. Added a form that allows the **logged in user** to submit a message (if you're logged out you will be taken to login page)
2. added a url for the messageboard - mainly for testing right now
3. in Views I have added the render function of the message form and the messages from the db, this function handles the POST of the messages
4. added board.html 
5. altered 005 test data file for the migrations to work (msgID needed to be changed to msg)

Each course page will have this model of messages so that is a task for the upcoming days (@DeanCivlin is close to finish it).

closes #109

image

### CODE REVIEW

File: AcadeMe/urls.py [hunk 1]
```diff
@@ -25,5 +25,7 @@
     path('register/', user_views.register, name='register'),
     path('login/', auth_views.LoginView.as_view(template_name='../templates/login/login.html'), name='login'),
     path('logout/', auth_views.LogoutView.as_view(template_name='users/logout.html'), name='logout'),
-    path('contact_us/', views.contact, name='contact_us')
+    path('contact_us/', views.contact, name='contact_us'),
+    path('msgboard/', user_views.msgboard, name='msgboard')
+
 ]
```
Comment (shakedhagag) [inline AcadeMe/urls.py:31]: The path to the msgboard link for testing

File: AcadeMeData/forms.py [hunk 2]
```diff
@@ -32,3 +32,9 @@ def save(self, commit=True):
         user2 = User(user=user, university=self.cleaned_data['university'], degree=self.cleaned_data['degree'])
         user2.save()
         return user2
+
+
+class MessageForm(forms.ModelForm):
+    class Meta:
+        model = Messages
+        fields = ('text', 'userID')
```
Comment (shakedhagag) [inline AcadeMeData/forms.py:40]: The form of Messages

File: AcadeMeData/migrations/0005_messageTags_test_data.py [hunk 3]
```diff
@@ -16,7 +16,7 @@ def generate_msgtagsdata(apps, schema_editor):
         ]
         with transaction.atomic():
             for id, msgID, userID in test_data:
-                msgTag = MessageTags(id=id, msgID=msgID, userID=userID)
+                msgTag = MessageTags(id=id, msg=msgID, userID=userID)
```
Comment (shakedhagag) [inline AcadeMeData/migrations/0005_messageTags_test_data.py:19]: The little change that was needed for the migrations to work again

File: AcadeMeData/views.py [hunk 4]
```diff
@@ -1,6 +1,8 @@
 from django.shortcuts import render, redirect
 from django.contrib import messages
-from .forms import UserRegistrationForm
+from .forms import UserRegistrationForm, MessageForm
+from .models import Messages
+from django.contrib.auth.decorators import login_required
```
Comment (shakedhagag) [inline AcadeMeData/views.py:6]: login required handles the auth of the user 
if youre logged out you will be redirected to login page

**Review by DeanCivlin — APPROVED** [review]
LGTM

**Review by DanWaldman — APPROVED** [review]
Looks good, in future PR we will be linking each board to specific course, nice work

**Review by razashtar93 — APPROVED** [review]
LGTM

**Review by DanielItkes — APPROVED** [review]

**Comment by Yarboa** [comment]
Nice work
Unfortunately there are missing tests. 
Please refer this https://github.com/redhat-beyond/AcadeMe/blob/main/contributing.md#guidelines-for-pr-reviews

**Comment by shakedhagag** [comment]
> Nice work Unfortunately there are missing tests. Please refer this https://github.com/redhat-beyond/AcadeMe/blob/main/contributing.md#guidelines-for-pr-reviews

Done! added the test file for the feature
please review 🍔 
@Yarboa  🥲

File: AcadeMe/urls.py [hunk 5]
```diff
@@ -25,5 +25,7 @@
     path('register/', user_views.register, name='register'),
     path('login/', auth_views.LoginView.as_view(template_name='../templates/login/login.html'), name='login'),
     path('logout/', auth_views.LogoutView.as_view(template_name='users/logout.html'), name='logout'),
-    path('contact_us/', views.contact, name='contact_us')
+    path('contact_us/', views.contact, name='contact_us'),
+    path('msgboard/', user_views.msgboard, name='msgboard')
+
```
Comment (Omeramsc) [inline AcadeMe/urls.py:30]: no need for the empty line

File: AcadeMeData/tests/test_msgboard_feature.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import pytest
+from AcadeMeData.models import User
+from AcadeMeData.forms import MessageForm
+
+
+@pytest.fixture
+def user_example():
```
Comment (Omeramsc) [inline AcadeMeData/tests/test_msgboard_feature.py:7]: Any reason not to use `user_example` from `test_register` or from `test_user`? you should have all of those consolidated into a contest instead of rewriting them for each and every test module.

**Review by Omeramsc — CHANGES_REQUESTED** [review]
Left some comments you should address, 
BTW I appreciate the comments you added along the code, but if you feel like you need comments to explain what had been done on each stage, it is usually a sign the code is not readable enough or requires changes. 
I did however like the PR description, very helpful!

File: AcadeMeData/tests/test_msgboard_feature.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    return user
+
+
+@pytest.mark.django_db
+def test_msgboard_feature(client, user_example):
+    client.force_login(user_example.user)
+    response = client.get("/msgboard/")
+    assert response.status_code == 200
+    form = response.context["form"]
+    assert isinstance(form, MessageForm)
```
Comment (Omeramsc) [inline AcadeMeData/tests/test_msgboard_feature.py:21]: I'm not sure what this assertion is testing, can you explain?
anyway, you should check that a message inserted by a user appear in the messageboard.

File: AcadeMe/urls.py (hunk 5 above)
Comment (shakedhagag) [reply AcadeMe/urls.py:30]: got it will change

File: AcadeMeData/tests/test_msgboard_feature.py (hunk 6 above)
Comment (shakedhagag) [reply AcadeMeData/tests/test_msgboard_feature.py:7]: What do you mean? importing them or merging the tests in to one file?
If importing is the answer then I have tried and failed so I changed it to this way

File: AcadeMeData/tests/test_msgboard_feature.py (hunk 7 above)
Comment (shakedhagag) [reply AcadeMeData/tests/test_msgboard_feature.py:21]: Asserting if the form that is being filled in /msgboard/ page is an instance of MessageForm
which is the form that handles messages

BTW 
I have written comments on my code because past experience taught me that it would make the PR approved faster, I do believe it is well written

File: AcadeMeData/tests/test_msgboard_feature.py (hunk 6 above)
Comment (Yarboa) [reply AcadeMeData/tests/test_msgboard_feature.py:7]: Please refer this review also 
https://github.com/redhat-beyond/AcadeMe/pull/103#discussion_r775504679

If you do not understand, please try friends in slack. understand it and propose a code reply

**Review by Yarboa — COMMENTED** [review]
I think test coverage is not enough,
Please try to suggest more tests.

File: AcadeMeData/tests/test_msgboard_feature.py (hunk 6 above)
Comment (shakedhagag) [reply AcadeMeData/tests/test_msgboard_feature.py]: got it, fixed this in the recent push

**Comment by shakedhagag** [comment]
> I think test coverage is not enough, Please try to suggest more tests.

Added more tests and organized them accordingly
fixed confest and inserted more fixtures into it

File: AcadeMeData/tests/test_msgboard_feature.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+@pytest.mark.django_db
+def test_msgboard_empty_message_when_page_loads(client, user_example):
+    client.force_login(user_example.user)
+    response = client.get("/msgboard/")
+    messages = list(get_messages(response.wsgi_request))
+    assert len(messages) == 0
+
+
+@pytest.mark.django_db
+def test_msgboard_feature(client, user_example, generate_msgboard):
```
Comment (Yarboa) [inline AcadeMeData/tests/test_msgboard_feature.py:30]: I would add here validation of the following
      assert len(response.context['messages']) > 0
      and also search for the content

File: AcadeMeData/tests/test_msgboard_feature.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    response = client.get("/msgboard/")
+    form = response.context["form"]
+    assert isinstance(form, MessageForm)
+
+
+@pytest.mark.django_db
+def test_msgboard_empty_message_when_page_loads(client, user_example):
+    client.force_login(user_example.user)
+    response = client.get("/msgboard/")
+    messages = list(get_messages(response.wsgi_request))
```
Comment (Yarboa) [inline AcadeMeData/tests/test_msgboard_feature.py:25]: Consider this 
 len(response.context['messages'])  instead

**Comment by Yarboa** [comment]
@shakedhagag look at test which is failing, you have no messages

**Comment by shakedhagag** [comment]
> @shakedhagag look at test which is failing, you have no messages

Yeah I am still trying to figure out why POST is not working..
im on this since 4 pm ..

File: AcadeMeData/tests/test_msgboard_feature.py (hunk 9 above)
Comment (shakedhagag) [reply AcadeMeData/tests/test_msgboard_feature.py:25]: The post test is giving me a hard time, checking everything from a to z, it only fills the form on /msgboard/ but not posting it and adding to the db
I think this is something that we can deal with later on, everything works manually, posting messages to msgboard is working perfectly 
(messages are being added to the DB) and other tests are fine too.
In order to get ahead with the project I would love for this pr to be merged and promise to deal with this specific test later on.

Comment (Omeramsc) [reply AcadeMeData/tests/test_msgboard_feature.py:25]: Tests are the most important part of the code, more than the functionality itself; it is not something we usually push aside for later. 
your code may work now, but a future feature (even if not related) may break it without you even noticing.

File: AcadeMeData/tests/test_msgboard_feature.py (hunk 7 above)
Comment (Omeramsc) [reply AcadeMeData/tests/test_msgboard_feature.py:18]: There's no use-case in which the form should not be a MessageForm, so I'm not sure why this assertion is needed. 
A better test would check the fields of the form we expect to see, or that it is saved correctly.

**Comment by Yarboa** [comment]
It can not be reviewed, please resolve conflicts and push,

File: AcadeMeData/tests/test_msgboard_feature.py (hunk 9 above)
Comment (shakedhagag) [reply AcadeMeData/tests/test_msgboard_feature.py:25]: I understand and agree, but Its taking me to much time to solve it,
I have read the docs, countless stackoverflow questions
I have consulted mentors, friends, different teams and no one could help 🥲 
If you want I can show you what the problem is via zoom

Comment (Omeramsc) [reply AcadeMeData/tests/test_msgboard_feature.py:25]: I really do get your frustration. but it does not meet our definition of done. 
sometimes it is better not having a feature than having something that does not meet the standards we decided on (good coverage of tests in this situation). 
I am glad to hear it made you consult others, as this is a good way to learn. 

I would've liked to see the problem, but I don't think I'll be available for this before the demo...

Comment (shakedhagag) [reply AcadeMeData/tests/test_msgboard_feature.py:25]: I'll try again 🥲 
thx

**Comment by Yarboa** [comment]
I think this PR should wait for #116

**Comment by shakedhagag** [comment]
rebased and pushed

File: AcadeMeData/tests/test_msgboard_feature.py (hunk 8 above)
Comment (Yarboa) [reply AcadeMeData/tests/test_msgboard_feature.py:33]: sorry you are avoiding the reviews, again, check that you have returned messages, you should not check the form
Not instance, content

**Review by Yarboa — COMMENTED** [review]
Try to be more careful and review the comments
clients tests should check the responses empty and not empty
While forms could be tested under the views

it is useless to check instances, if data is not created due to a bug FYI

**Comment by shakedhagag** [comment]
> Try to be more careful and review the comments clients tests should check the responses empty and not empty While forms could be tested under the views
> 
> it is useless to check instances, if data is not created due to a bug FYI

My bad I deleted it few days ago from the repo and after rebase it has been deleted
Im still working on it locally

File: AcadeMe/views.py [hunk 10]
```diff
@@ -7,6 +7,7 @@ def app_layout(request):
 
 
 def homePage(request):
+    print(request.user)
```
Comment (razashtar93) [inline AcadeMe/views.py:10]: do we need this print?

Comment (shakedhagag) [reply AcadeMe/views.py:10]: no hell we dont!
ill delete it now sorry mate

**Review by razashtar93 — APPROVED** [review]
Looks great!

File: AcadeMeData/tests/test_msgboard_feature.py (hunk 8 above)
Comment (Yarboa) [reply AcadeMeData/tests/test_msgboard_feature.py:33]: nvm, at least test were added

**Review by Yarboa — APPROVED** [review]
LGTM, although i expected for more related to tests views against client experience
